test/jdk/java/lang/ref/FinalizerHistogramTest.java

```diff
@@ -32,37 +33,50 @@
  * @test
  * @summary Unit test for FinalizerHistogram
  * @modules java.base/java.lang.ref:open
- * @run main FinalizerHistogramTest
+ * @library /test/lib
+ * @build jdk.test.whitebox.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
+ * @run main/othervm
+ *      -Xbootclasspath/a:.
+ *      -XX:+UnlockDiagnosticVMOptions
+ *      -XX:+WhiteBoxAPI
+ *      FinalizerHistogramTest
  */
 
 public class FinalizerHistogramTest {
     static ReentrantLock lock = new ReentrantLock();
-    static volatile int wasInitialized = 0;
-    static volatile int wasTrapped = 0;
-    static final int objectsCount = 1000;
+    static final AtomicInteger initializedCount = new AtomicInteger(0);
+    static final AtomicInteger trappedCount = new AtomicInteger(0);
+    static final int OBJECTS_COUNT = 1000;
+
+    static WhiteBox wb;
+    static boolean refProResult;
 
     static class MyObject {
         public MyObject() {
             // Make sure object allocation/deallocation is not optimized out
-            wasInitialized += 1;
+            initializedCount.incrementAndGet();
         }
 
         protected void finalize() {
             // Trap the object in a finalization queue
-            wasTrapped += 1;
+            trappedCount.incrementAndGet();
             lock.lock();
         }
     }
 
-    public static void main(String[] argvs) {
+    public static void main(String[] argvs) throws InterruptedException {
         try {
             lock.lock();
-            for(int i = 0; i < objectsCount; ++i) {
+            for(int i = 0; i < OBJECTS_COUNT; ++i) {
                 new MyObject();
             }
-            System.out.println("Objects intialized: " + objectsCount);
-            System.gc();
-            while(wasTrapped < 1);
+            System.out.println("Objects intialized: " + initializedCount.get());
+            wb = WhiteBox.getWhiteBox();
+            wb.fullGC();
+            refProResult = wb.waitForReferenceProcessing();
```

Comment:
WhiteBox::waitForReferenceProcessing returns true if some progress has been made, not if all the 
processing is done.  Probably need a loop here, e.g.
```
while (wb.waitForReferenceProcessing()) {}
// All finalizable objects from prior gcs have made it to the finalizer queue.
```

Reply:
I've only ever seen it return `false`. But that won't necessarily always be true; I've added a loop.